pkg/kubelet/server/auth.go

```diff
@@ -81,35 +83,60 @@ func (n nodeAuthorizerAttributesGetter) GetRequestAttributes(u user.Info, r *htt
 
 	requestPath := r.URL.Path
 
-	// Default attributes mirror the API attributes that would allow this access to the kubelet API
-	attrs := authorizer.AttributesRecord{
-		User:            u,
-		Verb:            apiVerb,
-		Namespace:       "",
-		APIGroup:        "",
-		APIVersion:      "v1",
-		Resource:        "nodes",
-		Subresource:     "proxy",
-		Name:            string(n.nodeName),
-		ResourceRequest: true,
-		Path:            requestPath,
-	}
-
-	// Override subresource for specific paths
-	// This allows subdividing access to the kubelet API
+	fineGrained := utilfeature.DefaultFeatureGate.Enabled(features.KubeletFineGrainedAuthz)
+	var attrs []authorizer.Attributes
 	switch {
 	case isSubpath(requestPath, statsPath):
-		attrs.Subresource = "stats"
+		attrs = append(attrs, attributes(u, apiVerb, "stats", string(n.nodeName), requestPath))
 	case isSubpath(requestPath, metricsPath):
-		attrs.Subresource = "metrics"
+		attrs = append(attrs, attributes(u, apiVerb, "metrics", string(n.nodeName), requestPath))
 	case isSubpath(requestPath, logsPath):
 		// "log" to match other log subresources (pods/log, etc)
-		attrs.Subresource = "log"
+		attrs = append(attrs, attributes(u, apiVerb, "log", string(n.nodeName), requestPath))
 	case isSubpath(requestPath, checkpointPath):
-		attrs.Subresource = "checkpoint"
+		attrs = append(attrs, attributes(u, apiVerb, "checkpoint", string(n.nodeName), requestPath))
+	case isSubpath(requestPath, podsPath):
+		if fineGrained {
+			attrs = append(attrs, attributes(u, apiVerb, "pods", string(n.nodeName), requestPath))
+		}
+		attrs = append(attrs, attributes(u, apiVerb, "proxy", string(n.nodeName), requestPath))
+	case isSubpath(requestPath, runningPodsPath):
+		if fineGrained {
+			attrs = append(attrs, attributes(u, apiVerb, "pods", string(n.nodeName), requestPath))
+		}
+		attrs = append(attrs, attributes(u, apiVerb, "proxy", string(n.nodeName), requestPath))
+	case isSubpath(requestPath, healthzPath):
+		if fineGrained {
+			attrs = append(attrs, attributes(u, apiVerb, "healthz", string(n.nodeName), requestPath))
+		}
+		attrs = append(attrs, attributes(u, apiVerb, "proxy", string(n.nodeName), requestPath))
+	case isSubpath(requestPath, configzPath):
+		if fineGrained {
+			attrs = append(attrs, attributes(u, apiVerb, "configz", string(n.nodeName), requestPath))
+		}
+		attrs = append(attrs, attributes(u, apiVerb, "proxy", string(n.nodeName), requestPath))
```

Comment:
optional nit: I think it's a bit cleaner to put these cases in a separate switch statement before, and let the default case in the second switch add the proxy subresource.

```
if fineGrained {
  switch {
    	case isSubpath(requestPath, podsPath):
			attrs = append(attrs, attributes(u, apiVerb, "pods", string(n.nodeName), requestPath))
	case isSubpath(requestPath, runningPodsPath):
			attrs = append(attrs, attributes(u, apiVerb, "pods", string(n.nodeName), requestPath))
	case isSubpath(requestPath, healthzPath):
			attrs = append(attrs, attributes(u, apiVerb, "healthz", string(n.nodeName), requestPath))
	case isSubpath(requestPath, configzPath):
			attrs = append(attrs, attributes(u, apiVerb, "configz", string(n.nodeName), requestPath))
  }
}
```

Reply:
Good call. Done!